dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -1045,6 +1047,33 @@ public Knowledge visitVirtualLocation(Relation rel) {
             }
             return new Knowledge(must, new HashSet<>(must));
         }
+
+        @Override
+        public Knowledge visitSyncWith(Relation rel) {
+            Set<Tuple> must = new HashSet<>();
+            List<Event> events = new ArrayList<>();
+            events.addAll(program.getThreadEvents(Load.class));
+            events.addAll(program.getThreadEvents(Store.class));
+            events.addAll(program.getThreadEvents(Fence.class));
+            events.addAll(program.getThreadEvents(RMWXchgBase.class));
+            events.removeIf(e -> e instanceof Init);
+            for (Event e1 : events) {
+                for (Event e2 : events) {
+                    Thread thread1 = e1.getThread();
+                    Thread thread2 = e2.getThread();
+                    if (thread1 == thread2 || thread1.optSyncSet.isEmpty() || thread2.optSyncSet.isEmpty()) {
+                        continue;
+                    }
+                    if (thread1.optSyncSet.get().contains(thread2) && !exec.areMutuallyExclusive(e1, e2)) {
+                        must.add(new Tuple(e1, e2));
+                    }
+                    if (thread2.optSyncSet.get().contains(thread1) && !exec.areMutuallyExclusive(e1, e2)) {
+                        must.add(new Tuple(e1, e2));
+                    }
+                }
+            }
+            return new Knowledge(must, new HashSet<>(must));
```

Comment:
`must` is already a `Set`. Can't you simple do `return new Knowledge(must, must);`?

Reply:
Updated

Reply:
You should be careful with `return new Knowledge(must, must);` because sharing the same set instance for both may and must se is dangerous. Changing one of them (e.g., as part of the analysis) will implicitly change the other as well (cause they are the same instance). It is unlikely to make a problem for static relations, but if it ever does, it is gonna be painful to debug this!

Reply:
Good point. I did not think about this

Reply:
Thanks, changed back to use two different instances.